neat_ml/utils/figure_utils.py

```diff
@@ -0,0 +1,376 @@
+                     component.
+        y_col (str): Name of the column for the y-axis 
+                     component.
+        phase_col (str): Name of the column containing 
+                         phase information.
+        ax (Axes): matplotlib axis for plotting
+        xrange (list[int]): X-axis composition range.
+        yrange (list[int]): Y-axis composition range. 
+        n_components (int): Number of GMM components.
+        random_state (int): Seed for reproducibility.
+        region_colors (Optional[list[str]]): Colors for the 
+                                             phase regions.
+        boundary_color (str): Color for the boundary line.
+        resolution (int): Grid resolution for the contour plot.
+        decision_alpha (float): Opacity of the filled regions.
+        plot_regions (bool): Whether to plot the filled regions.
+
+    Returns:
+    --------
+        gmm, std_labels, boundary_points (tuple): 
+               Output tuple containing the GMM model, cluster labels, 
+               boundary points, and a list of Plotly contour
+               traces.
+    """
+    df_local, x_col = rename_df_columns(df, x_col)
+    x_features = df_local[phase_col].to_numpy().reshape(-1, 1)
+
+    gmm = GaussianMixture(n_components=n_components, random_state=random_state)
+    raw_labels = gmm.fit_predict(x_features)
+    x_comp = df_local[[x_col, y_col]].to_numpy()
+    std_labels, label_map = _standardise_labels(raw_labels, x_comp)
+    wrapper = GMMWrapper(gmm, x_comp, x_features)
+    x_start, x_stop = map(float, xrange)
+    y_start, y_stop = map(float, yrange)
+    xs = np.linspace(x_start, x_stop, resolution)
+    ys = np.linspace(y_start, y_stop, resolution)
+    xx, yy = np.meshgrid(xs, ys)
+    z_raw = wrapper.predict(np.c_[xx.ravel(), yy.ravel()])
+    z = np.vectorize(label_map.get)(z_raw).reshape(xx.shape)
+
+    if region_colors is None:
+        region_colors = ["lightsteelblue", "aquamarine"]
+
+    if plot_regions:
+        ax.contourf(
+            xx,
+            yy,
+            z,
+            levels=[-0.5, 0.5, 1.5],
+            colors=region_colors,
+            alpha=decision_alpha,
+        )
+
+    ax.contour(
+        xx,
+        yy,
+        z,
+        levels=[0.5],
+        colors=boundary_color,
+        linewidths=3,
+    )
+    proxy_artist = Line2D(
+        [0], 
+        [0], 
+        label="Decision Boundary", 
+        color=boundary_color, 
+        linewidth=3
+    )
+    ax.legend(handles=[proxy_artist])
+
+    boundary_points = extract_boundary_from_contour(z, xs, ys, level=0.5)
+    return gmm, std_labels, boundary_points
+
+def plot_gmm_composition_phase(
+    df: pd.DataFrame,
+    x_col: str,
+    y_col: str,
+    phase_col: str,
+    ax: Axes,
+    point_cmap: Optional[list[str]] = None,
+) -> None:
+    """
+    Trains a GMM and creates a scatter trace for data points.
+
+    Optionally creates region traces as well, mimicking the 
+    original script's logic.
```

Comment:
what is the "original script?" that doesn't mean much to me, and probably less to another reader?